sdk/dotnet/Pulumi/Core/Alias.cs

```diff
@@ -0,0 +1,80 @@
+﻿// Copyright 2016-2018, Pulumi Corporation
+
+#nullable enable
+
+namespace Pulumi
+{
+    /// <summary>
+    /// Alias is a partial description of prior named used for a resource. It can be processed in
+    /// the context of a resource creation to determine what the full aliased URN would be.
+    ///
+    /// Note there is a semantic difference between properties being absent from this type and properties
+    /// having the <see langword="null"/> value.Specifically, there is a difference between:
+    ///
+    /// <c>
+    /// new Alias { Name = "foo", Parent = null } // and
+    /// new Alias { Name = "foo" }
+    /// </c>
+    ///
+    /// The presence of a property indicates if its value should be used. If absent, then the value
+    /// is not used. So, in the above while <c>alias.Parent</c> is <see langword="null"/> for both,
+    /// the first alias means "the original urn had no parent" while the second alias means "use the
+    /// current parent".
+    ///
+    /// Note: to indicate that a resource was previously parented by the root stack, it is
+    /// recommended that you use:
+    ///
+    /// <c>Aliases = { new Alias { Parent = Pulumi.Stack.Root } }</c>
+    ///
+    /// This form is self-descriptive and makes the intent clearer than using:
+    ///
+    /// <c>Aliases = { new Alias { Parent = null } }</c>
+    /// </summary>
+    public class Alias

```

Comment:
Again, may be controversial, but should we make this `sealed` by default (at least for now)?

Reply:
Have sealed this.  I don't htink it's controversial. :)